proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+```swift
+extension Span {
+  func extracting(_ bounds: Range<Int>) -> dependsOn(self) Self
+  func extracting(_ bounds: some RangeExpression<Int>) -> dependsOn(self) Self
+  func extracting(_: UnboundedRange) -> dependsOn(self) Self
+}
+```
+
+##### Unchecked access to elements and subranges of elements:
+
+The `subscript` and the `extracting()` functions mentioned above all have always-on bounds checking of their parameters, in order to prevent out-of-bounds accesses. We also want to provide unchecked variants as an alternative for cases where bounds-checking is proving costly.
+
+```swift
+extension Span {
+  // Unchecked subscripting and extraction
+
+  /// Accesses the element at the specified `position`.
+  ///
+  /// This subscript does not validate `position`; this is an unsafe operation.
+  ///
+  /// - Parameter position: The position of the element to access. `position`
+  ///     must be a valid index that is not equal to the `endIndex` property.
+  public subscript(
+    unchecked position: Index
+  ) -> dependsOn(self) Element { get }
+
+  /// Accesses a contiguous subrange of the elements represented by this `Span`
+  ///
+  /// This subscript does not validate `bounds`; this is an unsafe operation.
+  ///
+  /// - Parameter bounds: A range of the collection's indices. The bounds of
+  ///     the range must be valid indices of the collection.
+  public extracting(
+    uncheckedBounds bounds: Range<Index>
+  ) -> dependsOn(self) Self
+
+  /// Accesses the contiguous subrange of the elements represented by
+  /// this `Span`, specified by a range expression.
+  ///
+  /// This subscript does not validate `bounds`; this is an unsafe operation.
+  ///
+  /// - Parameter bounds: A range of the collection's indices. The bounds of
+  ///     the range must be valid indices of the collection.
+  public subscript(
+    uncheckedBounds bounds: some RangeExpression<Index>
+  ) -> dependsOn(self) Span<Element>
+}
+```
+
+##### Index validation utilities:
+
+Every time `Span` uses a position parameter, it checks for its validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension Span {
+  /// Traps if `position` is not a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - position: an position to validate
+  public boundsCheckPrecondition(_ position: Index)
+
+  /// Traps if `bounds` is not a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - position: a range of positions to validate
+  public boundsCheckPrecondition(_ bounds: Range<Index>)
+}
+```
```

Comment:
These talk about `Index`, but IIRC we don't define `Index`.

Reply:
Found and replaced the remaining instances of the word Index

Reply:
We should define `Index`! As in, we should literally have `typealias Index = Int` in `Span`.

(SE-0437 has sunk the concept of `Index` down to the `~Copyable` world for the UnsafeBufferPointer types.)